### PR TITLE
Added plurals and context support to Translation

### PR DESCRIPTION
**Pull request by SkyLucilfer**
_Context: Translation/i18n_

Added plurals and context support for translation in project and Editor.

### Added API
- For project:
tr_n(message, plural_message, n, context = "")

- For Editor:
TTRN(message, plural_message, n, context = "")
DTRN(message, plural_message, n, context = "")
RTRN(message, plural_message, n, context = "")

- All existing translation functions can now add context with the last argument too.
tr(message, context = "") 
TTR(message, context = "") 
etc.

### Example usage

- Specifying with context:
tr("Connect", "Device")
tr("Connect", "Signal")

- Plurals translation:
print(tr_n("%d user likes this.", "%d users like this.", n) % n);

- If we want more precise control over the replacement of n in the string during plurals translation:
print(tr_n("One user likes this.", "{num} users like this.", n).format([n], "{num}"));

- In Editor
Some examples from changes in scene_tree_editor.cpp 

multiple_connections

one_connect_multi_groups_en

one_connect_multi_group_french

### Test project
[po_parsing_pot_gen_2.zip](https://github.com/godotengine/godot/files/4963116/po_parsing_pot_gen_2.zip)
PO files which I have used to test the updated PO parser are contained in the folder "PO files". Also did some tests using the new tr(), tr_n() interfaces in GDScript.
Also tested the POT generation to generate POT files containing msgctxt and msgid_plurals, both for projects and Editor.

### Performance testing [07/08/2020]
[performance_test.zip](https://github.com/godotengine/godot/files/5041681/performance_test.zip)
I have abstracted PO to a specified class, TranslationPO. So translation using CSV will have the same performance as last time, as it uses the same Translation class before this PR.
When we import from a PO file, the imported locale will use the TranslationPO class. For 14000 translations in the dictionary:

- tr() takes 54 ms to translate 10 000 strings ~ 0.005 ms to translate one string. 

- tr_n() takes 
  - 280 ms to translate 10 000 strings ~ 0.028 ms for one string with short plural-rule, like French, German etc.
  - 1250ms to translate 10 000 strings ~ 0.25 ms for one string with very long plural-rule, Arabic

For reference, for a game running at 120 FPS, it takes 8.33 ms to render one frame.
So tr_n() is quite slow, and should be used with care. However, I do cache the last tr_n() result so if a translation query is the same (same key, n and context), it will return very fast. I imagine multiple same queries can happen if an UI element uses tr_n().
I have tried my best to optimize tr_n() - by caching variables, converting recursion to iteration, but it doesn't improve much. The slow runtime comes from the cost of evaluating the plural expression. Right now I use the Expression class. If runtime is really a concern, maybe we could make it faster by writing a dedicated function to evaluate the plural expression without using the Expression class, or hardcoding all possible plural rules and match locale with it.

### Performance testing [27/08/2020]
With the PR #41519, the performance problem of tr_n() is fixed. This is done by having a dedicated file containing all the plural rules.
tr_n() now has roughly the same runtime as tr(), with extra 0.001 ms to translate one string compared to tr().

### CODE REVIEW

**Comment by pycbouh** [comment]
This approach for plurals cannot be reused across languages with rules different from English, can it?

**Comment by SkyLucilfer** [comment]
> This approach for plurals cannot be reused across languages with rules different from English, can it?

Can, it's based on the GNU gettext idea for translating plurals strings. The "n" passed in by the users will be used to guide the system to fetch the correct translation, based on the plural rule of the selected language.

**Comment by name-here** [comment]
Don't some languages have more than one plural, depending on quantity?  Does this support that?

**Comment by SkyLucilfer** [comment]
@name-here Yes it supports that. You might want to refer to the answer I posted above.

**Comment by dalexeev** [comment]
It looks like the `tr_n` function is not very suitable if you are using identifier system:

```gdscript
tr_n("MY_ID", "", n) # Or `tr_n("MY_ID", "MY_ID", n)`?
```

<details>

<summary>From the docs</summary>

There are two approaches to generate multilingual language games and applications. Both are based on a key:value system. The first is to use one of the languages as the key (usually English), the second is to use a specific identifier. <...> In general, games use the second approach and a unique ID is used for each string.

</details>

**Comment by Calinou** [comment]
@dalexeev In my experience, gettext PO files are heavily centered around using English text as identifiers. On the other hand, custom formats (like Godot's CSV format) and XLIFF tend to recommend using keys as identifiers.

**Comment by pycbouh** [comment]
> In my experience, gettext PO files are heavily centered around using English text as identifiers.

This is definitely the intended way to use it by the creators for translating Linux, but the file format itself is not enforcing this as a rule in any way. If you use keys as identifiers, some tools may warn you that your translation language is English (POEdit does that, for one), but it's on the user to handle this. In this case the user being the engine.

**Review by YeldhamDev — APPROVED** [review]
LGTM

File: core/object.h
Comment (akien-mga) [inline core/object.h:726]: It's a bit weird to have this inconsistency in the API. I guess if we need to return `String` here, we should likely do the same with `tr()`.

I'm not fully up-to-date on how the String and StringName usage will impact GDScript and other languages, maybe @vnen could shed some light.

But it's true that a priori, it makes sense to me for those to return a plain String (StringName is great for comparison / usage as key, but here it will always end up converted to a String for display or formatting I guess).

File: core/translation.h
Comment (akien-mga) [inline core/translation.h:85]: I guess you could make this static in the `.cpp` and avoid the need to have the debug define in the header.

File: core/translation.cpp
Comment (akien-mga) [inline core/translation.cpp:880]: Yeah the purpose only seems to give an array of all messages. I'm not sure what's the use case is beside debugging, so I guess it's fine. Remember to remove those `////` comments before merge BTW :)

File: core/translation.cpp
Comment (akien-mga) [inline core/translation.cpp:962]:
```suggestion
void Translation::set_plural_rule(const String &p_plural_rule) {
	// Set plural_forms and plural_rule.
	// p_plural_rule passed in has the form "Plural-Forms: nplurals=2; plural=(n >= 2);".
```

I'm a bit concerned about having the gettext-specific plural handling directly in the base `Translation`, which is also used for CSV and could theoretically be used by more input formats in the future (though we have no plans to support more than PO and CSV).

Maybe this should be abstracted some more and the gettext-specific logic should be in a `TranslationPO` Resource?

WDYT @vnen @YeldhamDev?

File: core/translation.cpp
Comment (akien-mga) [inline core/translation.cpp:1247]: Nice cleanup :)

File: core/translation.h
Comment (akien-mga) [inline core/translation.h:61]: Maybe we can find clearer names to disambiguate, instead of relying on overloading? They both seem to do quite different things.

File: core/translation.h
Comment (akien-mga) [inline core/translation.h:53]: It would be nice to try to measure the performance impact of using a more complex data structure on a somewhat msg-heavy test project, both for PO files and CSV files.

That would help validate that this is the right approach, compared to e.g. using separate `Translation` resources for each context for example.

File: core/ustring.h
Comment (akien-mga) [inline core/ustring.h:415]: You forgot to define the `TTRN` stub for the `#else` branch, so this will fail compiling `tools=no` once we start using the feature.

File: core/ustring.cpp
Comment (akien-mga) [inline core/ustring.cpp:4347]:
```suggestion
	if (p_n > 1) {
		return p_text_plural;
	}
	return p_text_plural;
```
Might be a bit safer if we pass `-1` as `n` for example ;)

BTW, style wise you can simplify all tests like:
```cpp
if (a) {
	return something;
} else {
	return other_thing;
}
```
to
```cpp
if (a) {
	return something;
}

return other_thing;
```
(where `a` would typically be the most "special" condition and the last return would be the default return value)

File: editor/scene_tree_editor.cpp
Comment (akien-mga) [inline editor/scene_tree_editor.cpp:274]:
```suggestion
			msg_temp += "\n" + TTR("Click to show signals dock.");
```

File: doc/classes/EditorTranslationParserPlugin.xml
Comment (akien-mga) [inline doc/classes/EditorTranslationParserPlugin.xml:33]:
```suggestion
		msgids_context_plural.append(["Test 1", "context", "test 1 plurals"]) # This will add a message with msgid "Test 1", msgctxt "context", and msgid_plural "test 1 plurals".
		msgids_context_plural.append(["A test without context", "", "plurals"]) # This will add a message with msgid "A test without context" and msgid_plural "plurals".
		msgids_context_plural.append(["Only with context", "a friendly context", ""]) # This will add a message with msgid "Only with context" and msgctxt "a friendly context".
```
The line are also a bit long, I'd suggest putting the comments before each line of code.

File: editor/editor_translation_parser.cpp
Comment (akien-mga) [inline editor/editor_translation_parser.cpp:57]:
```suggestion
			ERR_FAIL_COND_V_MSG(arr.size() != 3, ERR_INVALID_DATA, "Array entries written into `msgids_context_plural` in `parse_file()` method should have the form [\"message\", \"context\", \"plural message\"]");
```
Removed the class name to make it shorter, and it should already be clear from the location of the error message I think.

Comment (akien-mga) [reply editor/editor_translation_parser.cpp:57]: You should add clearer documentation about this requirement in the XML doc itself, it's not something users should have to discover through an error (even though it's partly covered in the examples, but it's not explained that it's a hard requirement and what each argument means).

File: editor/pot_generator.h
Comment (akien-mga) [inline editor/pot_generator.h:57]: Same as in `translation.h`, I guess this could be kept only in the `.cpp` if you make it static.

File: editor/translations/extract.py
Comment (akien-mga) [inline editor/translations/extract.py:173]: Go ahead and add it, it shouldn't hurt :)

**Review by akien-mga — COMMENTED** [review]
Awesome work!

I left a few suggestions but it's mostly good to merge already.

My main concern is about adding so much gettext-specific logic to the core `Translation` (also used by `CSV`, and the abstract API that users access), instead of a more PO-specific resource type derived from `Translation`. I'd welcome input from other contributors on this.

File: core/io/translation_loader_po.cpp
Comment (SkyLucilfer) [reply core/io/translation_loader_po.cpp:169]: Yes, the maximum for x will be 5 (for plural in Arabic).

File: core/ustring.cpp
Comment (SkyLucilfer) [reply core/ustring.cpp:4347]: @akien-mga I wondered about negatives too. It turns out people usually use plural for negatives although there are no definite answers. See this [discussion](https://meta.stackexchange.com/questions/15590/are-negative-numbers-singular-or-plural). 0 is also written as plurals in english based on PO gettext rule. So I'll stick with the 1 check.

Comment (akien-mga) [reply core/ustring.cpp:4347]: Oh you're right, I didn't even think that negative was a valid input, but of course it is. It's good then.

File: editor/pot_generator.h
Comment (akien-mga) [reply editor/pot_generator.h:57]: Ah no, it's fine if it has to be this way.

**Comment by dalexeev** [comment]
> It looks like the `tr_n` function is not very suitable if you are using identifier system

> In my experience, gettext PO files are heavily centered around using English text as identifiers. On the other hand, custom formats (like Godot's CSV format) and XLIFF tend to recommend using keys as identifiers.

For CSV, we should also implement plurals support. For example like this:

KEY        |en           |ru
-----------|-------------|-------------
DAYS_AGO[0]|%d day ago   |%d день назад
DAYS_AGO[1]|%d days ago  |%d дня назад
DAYS_AGO[2]|-            |%d дней назад

Using:

```gdscript
var s = tr_n(n, "DAYS_AGO")
```

That is, we just have to make n the first argument, and it will be compatible with both systems.

Also, maybe we should remove the `_` from the function name? Rename to `trn` or `ntr`?

**Comment by SkyLucilfer** [comment]
Hello @dalexeev  
I talked to core devs about adding plurals for CSV too. However, it seems that the advantage of CSV is its simplicity, and trying to handle plurals with it would cancel out that benefit. In addition, let's say we need to support Arabic language in the CSV file, we will need 6 rows for every key, so it can be inconvenient. Therefore, in case we want plurals, we might be better off just use PO files. You could still make a proposal and see how the community responds to it. 
The reason I chose tr_n() is because I want to make it very explicit that we're translating plurals - tr stands for translate, and n indicates plurals. If there's no particular reason that the interface name is inconvenient or unusable, I would prefer not to change it.

**Comment by dalexeev** [comment]
@SkyLucilfer Thanks for the answer.

> However, it seems that the advantage of CSV is its simplicity, and trying to handle plurals with it would cancel out that benefit.

Is it difficult to implement or use? In my opinion, both are simple.

```gdscript
var s = tr("DAYS_AGO[%d]" % f(n)) % n
# where f is "Plural-Forms" logic. For English:
# func f(n: int) -> int:
#     if n == 1:
#         return 0
#     else:
#         return 1
```
> In addition, let's say we need to support Arabic language in the CSV file, we will need 6 rows **for every key**, so it can be inconvenient.

No, only for strings with number substitution. And there are usually few such strings:

KEY           |en         |ru
--------------|-----------|------------
REGULAR_KEY   |Regular key|Обычный ключ
...           |...        |...
SPECIAL_KEY[0]|%d key     |%d ключ
SPECIAL_KEY[1]|%d keys    |%d ключа
SPECIAL_KEY[2]|           |%d ключей
...           |...        |...
ANOTHER_KEY   |Another key|Другой ключ
...           |...        |...

> Therefore, in case we want plurals, we might be better off just use PO files.

That is, I will not be able to use the identifier system? In addition, it will be inconvenient to use a non-English language as the main one.

> You could still make a proposal and see how the community responds to it.

I'll add the proposal later.

**Comment by akien-mga** [comment]
For CSV plurals, I would suggest opening a proposal indeed and doing research on how plurals are handled by other projects that support CSV translations.

From what I found, there are many different CSV translation workflows and the few that support plurals have it hacked in in a way as suggested e.g. here, but there's no common standard. It's a simple system so we can indeed design our own plurals logic, but if there was a somewhat "popular" way of doing plurals with CSV used e.g. in other game engines, it would be best for us to follow that.

**Comment by dalexeev** [comment]
> If there's no particular reason that the interface name is inconvenient or unusable, I would prefer not to change it.

Maybe change at least the order of the arguments?

> ```gdscript
> # tr_n(message, plural_message, n, context = "")
> var s = tr_n("%d day ago", "%d days ago", n) % n
> ```
> 
> ```po
> # ru.po
> msgid "%d day ago"
> msgid_plural "%d days ago"
> msgstr[0] "%d день назад"
> msgstr[1] "%d дня назад"
> msgstr[2] "%d дней назад"
> ```
> 
> If using IDs:
> 
> ```gdscript
> # tr_n(message, plural_message, n, context = "")
> var s = tr_n("DAYS_AGO", "", n) % n
> ```
> 
> ```po
> # en.po
> msgid "DAYS_AGO"
> msgid_plural ""
> msgstr[0] "%d day ago"
> msgstr[1] "%d days ago"
> ```
> 
> I suggested changing the order of the arguments:
> 
> > That is, we just have to make `n` the first argument, and it will be compatible with both systems.
> 
> ```gdscript
> # tr_n(n, message, plural_message = "", context = "")
> var s = tr_n(n, "DAYS_AGO") % n
> ```

File: core/translation_po.cpp
Comment (akien-mga) [inline core/translation_po.cpp:31]: Add a line after this one (see https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes).

File: core/translation_po.cpp
Comment (akien-mga) [inline core/translation_po.cpp:32]: We tend to favor using absolute paths, so use `core/os/file_access.h`.

**Comment by akien-mga** [comment]
I've only read the last two commits diagonally, but overall this is very solid work so I think it's best to merge as is, and see if any follow ups are needed once using it in the `master` branch.

**Comment by akien-mga** [comment]
Thanks, and congrats for your top notch work on these features! :tada:

**Pull request merged**
